src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -0,0 +1,126 @@
+    auto values = pmr_vector<T>{alloc};
+    values.reserve(num_elements);
+    auto null_values = pmr_vector<bool>{alloc};
+    null_values.reserve(num_elements);
+
+    // copy values and null flags from value segment
+    auto iterable = ValueSegmentIterable<T>{*value_segment};
+    iterable.with_iterators([&](auto it, auto end) {
+      for (; it != end; ++it) {
+        auto segment_value = *it;
+        values.emplace_back(segment_value.value());
+        null_values.emplace_back(segment_value.is_null());
+      }
+    });
+
+    // LZ4 compression
+    const auto input_size = static_cast<int>(values.size() * sizeof(T));
+    auto output_size = LZ4_compressBound(input_size);
+    auto compressed_data = pmr_vector<char>{alloc};
+    compressed_data.resize(static_cast<size_t>(output_size));
+    const int compression_result = LZ4_compress_HC(reinterpret_cast<char*>(values.data()), compressed_data.data(),
+                                                   input_size, output_size, LZ4HC_CLEVEL_MAX);
+    Assert(compression_result > 0, "LZ4 compression failed");
+
+    auto data_ptr = std::allocate_shared<pmr_vector<char>>(alloc, std::move(compressed_data));
+    auto null_values_ptr = std::allocate_shared<pmr_vector<bool>>(alloc, std::move(null_values));
+
+    return std::allocate_shared<LZ4Segment<T>>(alloc, data_ptr, null_values_ptr, nullptr, compression_result,
+                                               input_size, num_elements);
+  }
+
+  std::shared_ptr<BaseEncodedSegment> _on_encode(
+      const std::shared_ptr<const ValueSegment<std::string>>& value_segment) {
+    const auto alloc = value_segment->values().get_allocator();
+    const auto num_elements = value_segment->size();
+    DebugAssert(num_elements <= std::numeric_limits<int>::max(),
+                "Trying to compress a ValueSegment with more "
+                "elements than fit into an int.");
+
+    // copy values and null flags from value segment
+    auto values = pmr_vector<char>{alloc};
+    auto null_values = pmr_vector<bool>{alloc};
+    null_values.reserve(num_elements);
+
+    /**
+     * These offsets mark the beginning of strings (and therefore end of the previous string) in the data vector.
+     * These offsets are character offsets. The string at position 0 starts at the offset stored at position 0.
+     * Its exclusive end is the offset stored at position 1 (i.e. offsets[1] - 1 is the last character of the string
+     * at position 0).
+     * In case of the last string its end is determined by the end of the data vector.
+     */
+    auto offsets = pmr_vector<size_t>{alloc};
+    offsets.reserve(num_elements);
+
+    auto iterable = ValueSegmentIterable<std::string>{*value_segment};
+    iterable.with_iterators([&](auto it, auto end) {
+      size_t offset = 0u;
+      bool is_null;
+      for (; it != end; ++it) {
+        auto segment_value = *it;
+        is_null = segment_value.is_null();
+        null_values.emplace_back(is_null);
```

Comment:
Yes this could be done in one line. But these variables are used in other places of the loop (e.g., the `is_null` variable). I don't think that minimizing the number of lines in this case will improve readability.

Reply:
Missed that. Nevermind.